<div class="alert alert-block alert-success">
    <h1 align="center">COVID-19 Vaccination Progress Analysis</h1>
    <h3 align="center">A Data Science Project using OWID + Manufacturer Datasets</h3>
    
<a href="https://www.kaggle.com/code/gpreda/covid-19-vaccination-progress">
Click for more information for dataset
    </a>
</div>

<img 
    src="https://kfor.com/wp-content/uploads/sites/3/2023/01/COVID19.jpg?w=960&h=540&crop=1"
    style="height:380px; object-fit:cover; width:100%; border-radius:12px;"
/>


<div style="background-color:#e6f9e6;border-left:6px solid #4caf50;padding:15px 20px;border-radius:8px;margin:15px 0;font-size:16px;line-height:1.5;">
    <p style="font-weight:bold;margin-bottom:10px;">
        Welcome to this notebook...!
    </p>
    <ul style="margin:0;padding-left:20px; text-align:justify;">
        <li>
            where we explore the global COVID-19 vaccination progress through a question-driven Exploratory Data Analysis (EDA). 
        In this study, we analyze vaccination trends across countries, uncover key patterns using data visualization, and identify a suitable country for forecasting future vaccination behavior. 
        The goal is to transform raw vaccination data into meaningful insights that support data-driven prediction and informed conclusions.
        </li>
    </ul>
</div>

<div style="background-color:#e6f9e6;border-left:6px solid #4caf50;padding:15px 20px;border-radius:8px;margin:15px 0;font-size:16px;line-height:1.5;">
    <p style="font-weight:bold;margin-bottom:10px;">
        Topics:
    </p>
    <ul style="margin:0;padding-left:20px;">
        <li>Import the Datasets (country_vaccinations - country_vaccinations_by_manufacturer)</li>
        <li>Exploratory Data Analysis (Question-driven EDA)</li>
        <li>Visualization</li>
        <li>Country Selection for Forecasting</li>
        <li>Prediction Insights and Final Conclusions</li>
    </ul>
</div>

<div style="
    background-color: #e6f9e6;
    border-left: 6px solid #4caf50;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Importing the libraries:
</div>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.io as pio
pio.renderers.default = "notebook_connected"


import warnings
warnings.filterwarnings('ignore')


plt.style.use('seaborn-v0_8')
sns.set_palette("husl")
pd.set_option('display.max_columns', None)

<div style="
    background-color: #e6f9e6;
    border-left: 6px solid #4caf50;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Load Datasets:
</div>

In [2]:
df1 = pd.read_csv(r'/kaggle/input/covid-world-vaccination-progress/country_vaccinations.csv')
df2 = pd.read_csv(r'/kaggle/input/covid-world-vaccination-progress/country_vaccinations_by_manufacturer.csv')

In [3]:
display(df1.sample(3))
print('-'*70)
display(df2.sample(3))

,country,iso_code,date,total_vaccinations,people_vaccinated,people_fully_vaccinated,daily_vaccinations_raw,daily_vaccinations,total_vaccinations_per_hundred,people_vaccinated_per_hundred,people_fully_vaccinated_per_hundred,daily_vaccinations_per_million,vaccines,source_name,source_website
29219,Germany,DEU,2021-08-29,101858347.0,54373205.0,50319663.0,85498.0,236231.0,121.40,64.81,59.98,2816.0,"Johnson&Johnson, Moderna, Novavax, Oxford/Astr...",Robert Koch Institut,https://impfdashboard.de/
84405,Vietnam,VNM,2021-08-09,9987587.0,8984300.0,1003287.0,581768.0,432627.0,10.17,9.15,1.02,4407.0,"Abdala, Moderna, Oxford/AstraZeneca, Pfizer/Bi...",World Health Organization,https://covid19.gov.vn/ngay-23-3-so-mac-moi-co...
73736,Spain,ESP,2022-02-17,92121948.0,NaN,38834215.0,99444.0,101437.0,197.07,NaN,83.08,2170.0,"Johnson&Johnson, Moderna, Oxford/AstraZeneca, ...",Ministry of Health,https://www.mscbs.gob.es/profesionales/saludPu...


----------------------------------------------------------------------


,location,date,vaccine,total_vaccinations
9789,France,2021-07-14,Oxford/AstraZeneca,7429304
29659,United States,2021-08-03,Johnson&Johnson,13479802
644,Argentina,2021-05-03,Moderna,4


<div style="
    background-color: #e6f9e6;
    border-left: 6px solid #4caf50;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Dataset Information & Data Types:
</div>

In [4]:
print("Main Dataset Info:")
df1.info()
print('-'*70)
print("\nManufacturer Dataset Info:")
df2.info()

Main Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86512 entries, 0 to 86511
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              86512 non-null  object 
 1   iso_code                             86512 non-null  object 
 2   date                                 86512 non-null  object 
 3   total_vaccinations                   43607 non-null  float64
 4   people_vaccinated                    41294 non-null  float64
 5   people_fully_vaccinated              38802 non-null  float64
 6   daily_vaccinations_raw               35362 non-null  float64
 7   daily_vaccinations                   86213 non-null  float64
 8   total_vaccinations_per_hundred       43607 non-null  float64
 9   people_vaccinated_per_hundred        41294 non-null  float64
 10  people_fully_vaccinated_per_hundred  38802 non-null  float64
 11  daily_vac

<div style="
    background-color: #e6f9e6;
    border-left: 6px solid #4caf50;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Stating Main Dataset:
</div>

<div style="
    background-color:#FFFFDD;
    border-left:6px solid #FFFE00;
    padding:15px 20px;
    border-radius:8px;
    margin:15px 0;
    font-size:14px;
    line-height:1.5;
">
    <p style="margin:0;font-weight:bold;">
        Convert 'date' to datetime and show only columns with missing values
    </p>
</div>

In [5]:
# Convert date to datetime 
df1['date'] = pd.to_datetime(df1['date'])

# Missing values overview
print("Missing Values by Column:")
missing = df1.isna().sum()
missing_percent = (missing / len(df1)) * 100
missing_table = pd.DataFrame({'Missing Count': missing, 'Percentage': missing_percent.round(2)})
display(missing_table[missing_table['Missing Count'] > 0].sort_values('Missing Count', ascending=False))

Missing Values by Column:


,Missing Count,Percentage
daily_vaccinations_raw,51150,59.12
people_fully_vaccinated,47710,55.15
people_fully_vaccinated_per_hundred,47710,55.15
people_vaccinated,45218,52.27
people_vaccinated_per_hundred,45218,52.27
total_vaccinations,42905,49.59
total_vaccinations_per_hundred,42905,49.59
daily_vaccinations,299,0.35
daily_vaccinations_per_million,299,0.35


<div style="
    background-color: #e6f9e6;
    border-left: 6px solid #4caf50;
    padding:15px 20px;
    border-radius:8px;
    margin:15px 0;
    font-size:14px;
    line-height:1.5;
">
    <p style="font-weight:bold;margin-bottom:10px;">
        Exploratory Data Analysis (EDA) – Question-Driven Approach
    </p>
    <p style="font-weight:bold;margin-bottom:10px;">
        <strong>From this point onward</strong>, we shift from raw data inspection to a structured, question-driven exploration of the global COVID-19 vaccination dataset.
    </p>
    <p style="font-weight:bold;margin-bottom:10px;">
     The goal of this EDA is threefold:
    </p>
    <ul style="margin:0;padding-left:20px;">
        <li>
            Understand the structure, coverage, and quality of the data
        </li>
        <li>
            Identify patterns, leaders, and anomalies in global vaccination efforts
        </li>
        <li>
          Make a data-driven decision on which country to select for the final forecasting model
        </li>
    </ul>
     <p style="font-weight:bold;margin-bottom:10px;">
     We will explore the data through a series of meaningful questions, each accompanied by visualization and insight.
    </p>
     <p style="font-weight:bold;margin-bottom:10px;">
     Let’s begin.
    </p>
</div>

In [6]:
df1.describe()

,date,total_vaccinations,people_vaccinated,people_fully_vaccinated,daily_vaccinations_raw,daily_vaccinations,total_vaccinations_per_hundred,people_vaccinated_per_hundred,people_fully_vaccinated_per_hundred,daily_vaccinations_per_million
count,86512,4.360700e+04,4.129400e+04,3.880200e+04,3.536200e+04,8.621300e+04,43607.000000,41294.000000,38802.000000,86213.000000
mean,2021-09-01 17:09:45.907157504,4.592964e+07,1.770508e+07,1.413830e+07,2.705996e+05,1.313055e+05,80.188543,40.927317,35.523243,3257.049157
min,2020-12-02 00:00:00,0.000000e+00,0.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000,0.000000,0.000000,0.000000
25%,2021-05-27 00:00:00,5.264100e+05,3.494642e+05,2.439622e+05,4.668000e+03,9.000000e+02,16.050000,11.370000,7.020000,636.000000
50%,2021-09-04 00:00:00,3.590096e+06,2.187310e+06,1.722140e+06,2.530900e+04,7.343000e+03,67.520000,41.435000,31.750000,2050.000000
75%,2021-12-12 00:00:00,1.701230e+07,9.152520e+06,7.559870e+06,1.234925e+05,4.409800e+04,132.735000,67.910000,62.080000,4682.000000
max,2022-03-29 00:00:00,3.263129e+09,1.275541e+09,1.240777e+09,2.474100e+07,2.242429e+07,345.370000,124.760000,122.370000,117497.000000
std,NaN,2.246004e+08,7.078731e+07,5.713920e+07,1.212427e+06,7.682388e+05,67.913577,29.290759,28.376252,3934.312440


In [7]:
print(df1.date.min(),'---',df1.date.max())

2020-12-02 00:00:00 --- 2022-03-29 00:00:00


<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q1:  How many countries are included and what is the time coverage?
</div>

In [8]:
main_countries = set(df1['country'].unique())
manu_countries = set(df2['location'].unique())

print(f"Countries in Main dataset: {len(main_countries)}")
print(f"Countries in Manufacturer dataset: {len(manu_countries)}")
print(f"Overlap (countries in both): {len(main_countries.intersection(manu_countries))}")

# List of countries that have disaggregated data
common_countries = sorted(main_countries.intersection(manu_countries))
print(f"\nCountries with manufacturer breakdown ({len(common_countries)}):")
print(common_countries)

Countries in Main dataset: 223
Countries in Manufacturer dataset: 43
Overlap (countries in both): 42

Countries with manufacturer breakdown (42):
['Argentina', 'Austria', 'Belgium', 'Bulgaria', 'Chile', 'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Ecuador', 'Estonia', 'Finland', 'France', 'Germany', 'Hong Kong', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Japan', 'Latvia', 'Liechtenstein', 'Lithuania', 'Luxembourg', 'Malta', 'Nepal', 'Netherlands', 'Norway', 'Peru', 'Poland', 'Portugal', 'Romania', 'Slovakia', 'Slovenia', 'South Africa', 'South Korea', 'Spain', 'Sweden', 'Switzerland', 'Ukraine', 'United States', 'Uruguay']


<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q2: Who are the global vaccination leaders? (Cumulative)

</div>

In [10]:
idx = df1.groupby('country')['date'].idxmax()
latest = df1.loc[idx]
latest=latest.sort_values(by='total_vaccinations', ascending=False).head(10)

fig = px.bar(latest, 
             x='total_vaccinations', 
             y='country', 
             orientation='h',
             title='Top 10 Countries by Total Vaccine Doses Administered',
             text='total_vaccinations',
             color='people_fully_vaccinated_per_hundred',
             color_continuous_scale='Viridis')
fig.update_layout(yaxis={'categoryorder':'total ascending'}, height=600)
fig.show()

<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q3:  Which countries have the highest vaccination rates per capita?

</div>

In [ ]:
per_capita = (df1.sort_values("date").groupby("country", as_index=False)\
              .last().sort_values("total_vaccinations_per_hundred", ascending=False).head(10))
px.bar(per_capita, x='country', y=['total_vaccinations_per_hundred',
       'people_fully_vaccinated_per_hundred'],
       title='Top 10 Countries by Doses per 100 People',
       barmode='group', height=600).show()

<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q4: What does the global daily vaccination trend look like?

</div>

In [ ]:
daily_global = df1.groupby('date')['daily_vaccinations'].sum().reset_index()

fig = px.line(daily_global, x='date', y='daily_vaccinations',
              title='Global Daily Vaccinations (All Countries Combined)',
              labels={'daily_vaccinations': 'Daily Doses'})
fig.add_vline(x=pd.to_datetime('2021-04-01'), line_dash="dash", line_color="red")
fig.add_annotation(x='2021-04-01', y=daily_global['daily_vaccinations'].max(),
                   text="Peak Period", showarrow=True)
fig.show()

<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q5: Which country had the fastest vaccination campaign? (Highest peak daily rate)

</div>

In [11]:
idx= df1.groupby('country')['daily_vaccinations'].idxmax()
top_10 =df1.loc[idx]
peak_daily= top_10.sort_values(by='date', ascending=True).head(10)

px.bar(peak_daily, x='country', y='daily_vaccinations',
       title='Highest Single-Day Vaccination Records by Country',
       text='date').update_xaxes(categoryorder='total descending').show()

<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q6: Is there a weekend effect in vaccination campaigns?

</div>

In [12]:
df1['day_of_week'] = df1['date'].dt.day_name()
weekend_effect = df1.groupby('day_of_week')['daily_vaccinations_per_million'].mean()

px.bar(x=weekend_effect.index, y=weekend_effect.values,
       title='Average Daily Vaccinations per Million by Day of Week (Global)',
       labels={'y': 'Doses per Million'}).show()

<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q7: Monthly vaccination speed across continents/regions?

</div>

In [13]:
top_regions = df1[df1['country'].isin(['United States','India','Brazil','Germany','United Kingdom','France','Italy','Japan','Canada','South Korea'])]

monthly = top_regions.groupby([pd.Grouper(key='date', freq='M'), 'country'])['daily_vaccinations'].mean().reset_index()

fig = px.line(monthly, x='date', y='daily_vaccinations', color='country',
              title='Monthly Average Daily Vaccinations – Selected Major Countries')
fig.show()

<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q8: Data Quality Check – Which countries have the most consistent daily reporting?

</div>

In [14]:
reporting_consistency = df1.groupby('country').agg({
    'date': 'nunique',
    'daily_vaccinations': lambda x: x.notna().sum()
}).reset_index()
reporting_consistency['reporting_ratio'] = reporting_consistency['daily_vaccinations'] / reporting_consistency['date']

best_reporters = reporting_consistency.nlargest(15, 'reporting_ratio')

print("Top 15 Countries by Data Reporting Consistency")
display(best_reporters)

Top 15 Countries by Data Reporting Consistency


,country,date,daily_vaccinations,reporting_ratio
149,Norway,482,481,0.997925
110,Latvia,480,479,0.997917
54,Denmark,476,475,0.997899
212,United States,471,470,0.997877
35,Canada,470,469,0.997872
41,China,470,469,0.997872
163,Russia,470,469,0.997872
97,Israel,466,465,0.997854
115,Liechtenstein,463,462,0.997840
161,Qatar,463,462,0.997840


<div style="
    background-color:#DCFBFC;
    border-left:6px solid #10B8BC;
    padding:15px 20px;
    border-radius:8px;
    margin:15px 0;
    font-size:16px;
    line-height:1.5;
">
    <p style="font-weight:bold;margin-bottom:10px;">
        Quick EDA on the Manufacturer-Specific Dataset:
    </p>
    <ul style="margin:0;padding-left:20px;">
        <li>
            Although our main modeling will rely on the comprehensive OWID dataset, the manufacturer breakdown provides valuable context for storytelling and data validation.
        </li>
    </ul>
</div>

In [15]:
# Convert date to datetime 
df2['date'] = pd.to_datetime(df2['date'])
print(df1.date.min(),'---',df1.date.max())
print(df2.date.min(),'---',df2.date.max())

2020-12-02 00:00:00 --- 2022-03-29 00:00:00
2020-12-04 00:00:00 --- 2022-03-30 00:00:00


<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q1: Which vaccines were used in each country? (Interactive Heatmap)
</div>

In [16]:
vaccine_matrix = df2.pivot_table(
    index='location',
    columns='vaccine',
    values='total_vaccinations',
    aggfunc='max',
    fill_value=0
) > 0


vaccine_matrix = vaccine_matrix.loc[vaccine_matrix.sum(axis=1).sort_values(ascending=False).index]


fig = px.imshow(
    vaccine_matrix.astype(int),
    text_auto=False,
    aspect="auto",
    color_continuous_scale='Blues',
    title='Vaccine Portfolio by Country (Blue = Used)',
    height=900
)
fig.update_layout(xaxis_title="Vaccine", yaxis_title="Country")
fig.show()


<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q2: Question 2: Top 10 Countries by Vaccine Diversity
</div>

In [17]:
country_vaccine_count = df2.groupby('location')['vaccine'].nunique().sort_values(ascending=False)

fig = px.bar(
    x=country_vaccine_count.head(10).index,
    y=country_vaccine_count.head(10).values,
    text=country_vaccine_count.head(10).values,
    title='Top 10 Countries by Number of Different Vaccines Used',
    labels={'x': 'Country', 'y': 'Number of Vaccines'}
)
fig.update_traces(textposition='outside')
fig.show()


<div style="
    background-color: #FFE1FF;
    border-left: 6px solid #660066;
    padding: 15px 20px;
    border-radius: 8px;
    margin: 15px 0;
    font-size: 16px;
    line-height: 1.5;    
">
Q3: Which vaccines dominate globally, and which countries contribute the most to each vaccine?
</div>

In [18]:

# aggregate total doses by vaccine and by (vaccine, country)
by_vaccine = df2.groupby('vaccine', as_index=False)['total_vaccinations'].sum()
by_vaccine_country = df2.groupby(['vaccine','location'], as_index=False)['total_vaccinations'].sum()

# optional: keep only top contributors per vaccine to avoid فوق‌العاده شلوغ شدن
# here we keep top 10 countries per vaccine (tweak n as you like)
n = 10
top_contributors = (
    by_vaccine_country
    .groupby('vaccine', group_keys=False)
    .apply(lambda g: g.nlargest(n, 'total_vaccinations'))
    .reset_index(drop=True)
)

# create a combined dataframe: top level = vaccine, second level = country contributions
# for sunburst we need each row to belong to a path; include the vaccine totals as rows too.
vaccine_rows = by_vaccine.rename(columns={'total_vaccinations':'doses'})
country_rows = top_contributors.rename(columns={'total_vaccinations':'doses'})

# add a helper column for root path (optional)
# Build a single dataframe with path ['vaccine','location']
sun_df = pd.concat([
    # vaccine-level rows: have vaccine but no location (location set to "")
    vaccine_rows.assign(location=""),
    # country-level rows: vaccine + location
    country_rows
], ignore_index=True)

# draw sunburst: first ring = vaccine (global share), second ring = country contributors
fig = px.sunburst(
    sun_df,
    path=['vaccine','location'],   # levels: vaccine -> country
    values='doses',
    title='Global Vaccine Market Share and Top Country Contributors',
    height=600
)
fig.update_traces(textinfo='label+percent entry')  # show percent of parent
fig.show()


<div style="
    background-color:#DCFBFC;
    border-left:6px solid #10B8BC;
    padding:15px 20px;
    border-radius:8px;
    margin:15px 0;
    font-size:16px;
    line-height:1.5;
">
    <p style="font-weight:bold;margin-bottom:10px;">
        Selecting the Top Country for Forecasting:
    </p>
    <ul style="margin:0;padding-left:20px;">
        <li>
            Based on data quality, reporting consistency, and vaccination coverage, <strong>Germany</strong> has been identified as the top country for building a predictive model of COVID-19 vaccination trends.
        </li>
    </ul>
</div>

In [19]:
#Number of reporting days in the main dataset (per country)
days_main = df1.groupby('country')['date'].nunique()

#Build a daily table from manufacturer dataset (sum values if multiple records exist per day)
manu_daily = df2.groupby(['location', 'date'], as_index=False)['total_vaccinations'].sum()
manu_days = manu_daily.groupby('location')['date'].nunique()

#Merge on (country, date) and compute mean relative error (no loops)
merged = df1[['country','date','total_vaccinations']].merge(
    manu_daily.rename(columns={'location':'country','total_vaccinations':'total_manu'}),
    on=['country','date'],
    how='inner'
)

error_percent = (
    merged.assign(
        rel_error=lambda d: (d['total_vaccinations'] - d['total_manu']).abs()
                            / d['total_vaccinations'].replace(0, np.nan)
    )
    .groupby('country')['rel_error']
    .mean() * 100
)

#Combine indicators into a DataFrame and compute the final score
ranking = pd.DataFrame({
    'days_main': days_main,
    'days_manu': manu_days
}).reindex(days_main.index)

ranking['days_manu'] = ranking['days_manu'].fillna(0).astype(int)
ranking['error_percent'] = error_percent.reindex(ranking.index)

# Weighted scoring formula
ranking['score'] = (
    0.5 * (ranking['days_main'] / ranking['days_main'].max()) +  # main reporting completeness
    0.3 * (ranking['days_manu'] > 50).astype(int) +              # manufacturer coverage bonus
    0.2 * (1 - ranking['error_percent'].fillna(100) / 100)        # error penalty
)

final_ranking = ranking.sort_values('score', ascending=False).round(4)

display(
    final_ranking.head(15)[['days_main','days_manu','error_percent','score']]
    .sort_values(by='error_percent')
)


,days_main,days_manu,error_percent,score
country,,,,
Germany,458.0,458,0.0000,0.9751
France,457.0,457,0.0000,0.9741
Argentina,456.0,456,0.0000,0.9730
Switzerland,463.0,463,0.0675,0.9802
Czechia,458.0,458,0.8274,0.9734
Belgium,454.0,64,1.0579,0.9688
Romania,457.0,456,1.3797,0.9713
Lithuania,458.0,64,1.5981,0.9719
Norway,482.0,64,1.6813,0.9966


<div style="
    background-color:#DCFBFC;
    border-left:6px solid #10B8BC;
    padding:15px 20px;
    border-radius:8px;
    margin:15px 0;
    font-size:14px;
    line-height:1.5;
">
    <p style="font-weight:bold;margin-bottom:10px;">
        Final Country Selection
    </p>
     <p style="font-weight:bold;margin-bottom:10px;">
        After a rigorous multi-criteria analysis combining both datasets, <strong>Germany</strong> was selected as the case study with the following unbeatable advantages:
    </p>
    <ul style="margin:0;padding-left:20px;">
        <li>
            458 days of reporting in the main dataset
        </li>
        <li>
            Full 458-day coverage in the manufacturer dataset
        </li>
        <li>
            <strong>Perfect 0.0000% discrepancy</strong> between the two sources → highest data quality
        </li>
        <li>
            Diverse vaccine portfolio (Pfizer/BioNTech, Moderna, AstraZeneca, Johnson&Johnson)
        </li>
        <li>
            Well-documented campaign phases (start, peak, booster waves)
        </li>
    </ul>
     <p style="font-weight:bold;margin-bottom:10px;">
       <strong>Selected country: Germany</strong>
    </p>
</div>

In [20]:
ger_manu = df2[df2['location'] == 'Germany'].copy()


ger_pivot = ger_manu.pivot_table(
    index='date',
    columns='vaccine',
    values='total_vaccinations',
    aggfunc='sum'
).ffill().fillna(0)

fig = px.area(
    ger_pivot,
    title='Germany Vaccination Campaign by Manufacturer – The Full Story',
    labels={'value': 'Cumulative Doses', 'date': 'Date'},
    height=600
)
fig.update_layout(
    xaxis_title="Date",
    yaxis_title="Cumulative Doses Administered",
    legend_title="Vaccine"
)
fig.show()


print("Key Storytelling Points from Germany:")
print("• Started with Pfizer/BioNTech (Dec 2020)")
print("• AstraZeneca arrived Feb 2021 → sharp increase")
print("• Moderna joined March 2021")
print("• Johnson&Johnson (single-dose) boosted summer 2021")
print("• Perfect data alignment with OWID dataset (0.00% error) → ideal for modeling")

Key Storytelling Points from Germany:
• Started with Pfizer/BioNTech (Dec 2020)
• AstraZeneca arrived Feb 2021 → sharp increase
• Moderna joined March 2021
• Johnson&Johnson (single-dose) boosted summer 2021
• Perfect data alignment with OWID dataset (0.00% error) → ideal for modeling


<div style="background-color:#e6f9e6;border-left:6px solid #4caf50;padding:18px 22px;border-radius:8px;margin:20px 0;font-size:16px;line-height:1.6;text-align:justify;">
    <ul style="margin:0;padding-left:20px;list-style-type:disc;">    
        <li style="margin-bottom:10px;">
            <strong>Next Steps — Feel Free to Fork and Continue the Modeling</strong>
        </li>     
        <li style="margin-bottom:10px;">
            Up to this point, the analysis identified <strong>Germany</strong> as the most reliable and complete country for building predictive models based on COVID-19 vaccination data.
        </li>    
        <li style="margin-bottom:10px;">
            Germany provides the highest data consistency, minimal aggregation errors, and the richest daily timeline, making it an excellent candidate for time-series modeling.
        </li>
        <li style="margin-bottom:10px;">
            To keep this notebook focused on data preprocessing, validation, and country selection, the next stage — modeling and forecasting — is intentionally left open.
        </li>
        <li style="margin-bottom:10px;">
             <strong>If you are interested in time-series forecasting (ARIMA, Prophet, LSTM, XGBoost, etc.), feel free to fork this notebook and build your own models on top of the cleaned German dataset.</strong>
        </li> 
        <li style="margin-bottom:10px;">
            This provides a great opportunity for:
            <ul style="margin-top:8px;padding-left:20px;list-style-type:circle;">
                <li>experimenting with different forecasting algorithms</li>
                <li>comparing short-term vs long-term vaccination trends</li>
                <li>exploring model interpretability and error analysis</li>
                <li>extending the pipeline with additional external features (mobility, cases, policy index, etc.)</li>
            </ul>
        </li>
        <li style="margin-top:12px;">
            I hope this notebook serves as a clean and well-prepared starting point for anyone who wants to dive deeper into predictive modeling on vaccination time-series data.
            <br><br>
            <strong>Happy modeling! </strong>
        </li>
    </ul>
</div>
